keps/sig-api-machinery/5080-ordered-namespace-deletion/README.md

```diff
@@ -0,0 +1,832 @@
+
+3. Maintain Predictability and Consistency: Provide a more deterministic deletion process to improve user confidence and debugging during namespace cleanup.
+
+4. Integrate with Existing Kubernetes Concepts: Build on the namespace deletion’s current architecture without introducing breaking changes to existing APIs or workflows.
+
+5. Be safe - don’t introduce unresolvable deadlocks.
+
+6. Make the most common dependency - workloads and the policies that govern them - safe by default for all types of policies, including CRDs, unless specifically opted out.
+
+
+### Non-Goals
+
+1. Reordering Deletion Across Namespaces: This design focuses on resource deletion within a single namespace. It does not attempt to enforce or prioritize deletion order across multiple namespaces.
+
+2. Introducing Custom Per-Resource Deletion Order: While the proposal aims for opinionated ordering, it does not cover fine-grained customization by end-users for specific resources or workloads.
+
+3. Guaranteeing Real-Time Enforcement: The proposal does not aim to guarantee real-time deletion of resources; the Kubernetes control plane’s reconciliation loop remains the underlying driver.
+
+4. Replacing Finalizers or Current Garbage Collection Mechanisms: The design does not intend to replace or bypass the existing Finalizer mechanism but works alongside it to enhance the resource cleanup process.
+
+5. Handling Non-Standard or External Resources: This design does not apply to external or non-Kubernetes resources managed outside the cluster (e.g., external databases, cloud resources).
+
+6. Global Enforcement of Security Policies: While the proposed changes improve security during deletion, it is not a substitute for broader, cluster-wide security policies or mechanisms.
+
+7. Implement a full graph: This proposal does not aim to model the dependency relationship between specific objects (instances) or between types as an arbitrary graph.
+
+
+## Proposal
+
+Introduce a way to specify priority based on resource type while deleting namespace. When the feature gate `OrderedNamespaceDeletion` is enabled,
+the resources associated with this namespace should be deleted in order.
+
+To specify the deletion order, the options would be:
+
+Option 1: have int value assigned for resources to indicate the deletion priority such like:
+
+```
+{Resource: "/pods", DeletionPriority: "5"}
+{Resource: "networking.k8s.io/networkpolicies", DeletionPriority: "-999"}                     
+{Resource: "apps.k8s.io/deployments", DeletionPriority: "10"}            
+...
+```
+0 would be the default DeletionPriority if not specified. And the resources which have deletion ordering concern would make sure to set DeletionPriority ahead of the resource it guards.
+
+Option 2: have the deletion priority bands defined instead of using numbers.
+
+To begin with, the deletion order bands would be introduced:
+- Workload Controllers
```

Comment:
Does it really matter if you delete workload controllers before workloads? I assume the intent is to stop them from recreating Pods, but once the Namespace has a DeletionTimestamp you can't create new resources in it, right?

Reply:
It probably doesn't REALLY matter, but it seems reasonable to have a band before workloads, which is before default.  That said, it could be YAGNI

Reply:
Maybe we could begin with something simply and add order band later only if it's absolutely necessary? I am open with either way